### PR TITLE
[FSK-75]: Add Create Family Mutation

### PR DESCRIPTION
**Pull request by mhmda-83**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -39,7 +39,8 @@
     "pactum": "3.2.2",
     "reflect-metadata": "0.1.13",
     "rimraf": "3.0.2",
-    "rxjs": "7.5.6"
+    "rxjs": "7.5.6",
+    "slug": "8.0.0"
```
Comment (pourdaavar) [inline package.json:43]: there is [toKebabCase](https://github.com/fullstacksjs/toolbox#tokebabcase) method in toolbox.

File: src/family/dto/create-family.input.ts [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import { Field, InputType } from '@nestjs/graphql';
+
+@InputType()
+export class CreateFamilyInput {
+  @Field()
+  name: string;
+}
```
Comment (pourdaavar) [inline src/family/dto/create-family.input.ts]: it's better moved into dto folder

`dto/input/create-family-input.dto.ts`

File: src/shared/graphqlRequest.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
...
+import * as pactum from 'pactum';
+
+export function graphqlRequest(query: string) {
+  return pactum
+    .spec()
+    .post('/graphql')
+    .withHeaders('Content-Type', 'application/json')
+    .withGraphQLQuery(query)
+    .expectStatus(HttpStatus.OK);
+}
```
Comment (pourdaavar) [inline src/shared/graphqlRequest.ts]: I think this file should be part of the e2e folder.

File: src/family/family.service.ts [hunk 4]
```diff
@@ -1,21 +1,59 @@
...
+        });
+      } catch (error: any) {
+        if (error?.code === 'P2002' && error?.meta?.target?.includes('slug')) {
+          counter++;
+        } else {
+          this.logger.error({ errorOnCreateFamily: error });
+          throw error;
+        }
+      }
+    }
```
Comment (pourdaavar) [inline src/family/family.service.ts:55]: what's the purpose here?

**Review by pourdaavar — CHANGES_REQUESTED** [review]

File: src/family/family.service.ts (hunk 4 above)
Comment (mhmda-83) [reply src/family/family.service.ts:55]: if slug is duplicate it shouldn't throw error, it should add a number at the end of it.


for example we have two family with the name: Ali Alavi
the slug will be ali-alavi for first family and for second family it'll be ali-alavi-1

File: package.json (hunk 1 above)
Comment (mhmda-83) [reply package.json:43]: slug package is not converting space to hyphen (-) only,
it converts Persian to english also.

File: src/family/dto/create-family.input.ts (hunk 2 above)
Comment (mhmda-83) [reply src/family/dto/create-family.input.ts]: Done

File: src/shared/graphqlRequest.ts (hunk 3 above)
Comment (mhmda-83) [reply src/shared/graphqlRequest.ts]: Done

File: src/family/dto/input/create-family.input.ts [hunk 5]
```diff
@@ -0,0 +1,7 @@
+import { Field, InputType } from '@nestjs/graphql';
+
+@InputType()
+export class CreateFamilyInput {
+  @Field()
```
Comment (MohammadForoutan) [inline src/family/dto/input/create-family.input.ts:5]: we can have validation in dto files with `(class-validator)`

File: src/family/family.resolver.ts [hunk 6]
```diff
@@ -16,4 +17,9 @@ export class FamilyResolver {
   findOne(@Args('id', { type: () => String }) id: string) {
     return this.familyService.findById(id);
   }
+
+  @Mutation(() => Family)
+  createFamily(@Args('data') createFamilyInput: CreateFamilyInput) {
```
Comment (MohammadForoutan) [inline src/family/family.resolver.ts]: I think `input` is a better name than `data` (what is your opinion?)
I have used `input` 
for consistency it's better to use one of them

File: src/family/family.service.ts (hunk 4 above)
Comment (MohammadForoutan) [reply src/family/family.service.ts:55]: it's good but I think it's good to have a function for this.(we need it later in other modules too)
also we need to format errors and not simply throwing them(we can do this in another pull request)

File: src/family/family.resolver.ts (hunk 6 above)
Comment (mhmda-83) [reply src/family/family.resolver.ts]: I agree with input too.

**Review by MohammadForoutan — CHANGES_REQUESTED** [review]
I add comment on files. check them.

File: src/family/family.service.spec.ts [hunk 7]
```diff
@@ -57,4 +58,31 @@ describe('FamilyService', () => {
...
+
+    it('should throw unrecognized error', async () => {
+      jest
+        .spyOn(prisma.family, 'create')
+        .mockRejectedValueOnce(new Error('Wrong Input'));
+
+      await expect(
+        service.create({ name: faker.name.fullName() }),
+      ).rejects.toThrow();
+    });
```
Comment (pourdaavar) [inline src/family/family.service.spec.ts:86]: I think it's better if there is another test case for the expected response too

File: src/family/family.service.ts (hunk 4 above)
Comment (hoseinprd) [reply src/family/family.service.ts:55]: I think there is a better way to do this.

File: src/family/family.service.ts [hunk 8]
```diff
@@ -1,21 +1,59 @@
...
+          extendedData.slug = `${familySlug}-${counter}`;
+        }
+
+        // eslint-disable-next-line no-await-in-loop
+        family = await this.prismaService.family.create({
+          data: extendedData,
+        });
+      } catch (error: any) {
+        if (error?.code === 'P2002' && error?.meta?.target?.includes('slug')) {
+          counter++;
```
Comment (pourdaavar) [inline src/family/family.service.ts:49]: It's not good practice to have a query call in the loop,
you can do it in a better way,
`slug: family-random-string-expected-slug`

**Pull request closed without merging**